tests/amor/utils_test.py

```diff
@@ -0,0 +1,248 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2024 Scipp contributors (https://github.com/scipp)
+# flake8: noqa: F403, F405
+
+import pytest
+import scipp as sc
+
+from ess.amor.utils import q_theta_figure, wavelength_theta_figure, wavelength_z_figure
+
+
+@pytest.fixture
+def da():
+    return sc.DataArray(
+        data=sc.ones(dims=('events',), shape=(10,)),
+        coords={
+            'wavelength': sc.linspace('events', 1, 12, 10, unit='angstrom'),
+            'theta': sc.linspace('events', 0, 1, 10, unit='deg'),
+            'Q': sc.linspace('events', 0, 0.3, 10, unit='1/angstrom'),
+            'z_index': sc.arange('events', 0, 10),
+        },
+    )
+
+
+@pytest.fixture
+def wavelength_bins():
+    return sc.linspace('wavelength', 1, 12, 3, unit='angstrom')
+
+
+@pytest.fixture
+def theta_bins():
+    return sc.linspace('theta', 0, 1, 3, unit='deg')
+
+
+@pytest.fixture
+def q_bins():
+    return sc.linspace('Q', 0, 0.3, 3, unit='1/angstrom')
+
+
+def test_wavelength_figure_table(da, wavelength_bins, theta_bins):
+    with pytest.raises(ValueError, match='binning provided'):
+        wavelength_theta_figure(da)
+
+    with pytest.raises(ValueError, match='binning provided'):
+        wavelength_theta_figure(da, wavelength_bins=wavelength_bins)
+
+    with pytest.raises(ValueError, match='binning provided'):
+        wavelength_theta_figure(da, theta_bins=theta_bins)
+
+    assert wavelength_theta_figure(
+        da, theta_bins=theta_bins, wavelength_bins=wavelength_bins
+    )
+
+
+def test_wavelength_figure_binned(da, wavelength_bins, theta_bins):
+    with pytest.raises(ValueError, match='binning provided'):
+        wavelength_theta_figure(da.bin(wavelength=3))
+
+    with pytest.raises(ValueError, match='binning provided'):
+        wavelength_theta_figure(da.bin(theta=3))
+
+    assert wavelength_theta_figure(da.bin(wavelength=3, theta=3))
+    assert wavelength_theta_figure(da.bin(wavelength=3), theta_bins=theta_bins)
+    assert wavelength_theta_figure(da.bin(theta=3), wavelength_bins=wavelength_bins)
```

Comment:
What happens if you do
`wavelength_theta_figure(da.bin(theta=3), wavelength_bins=wavelength_bins, theta_bins=theta_bins)`
?

Generally, is it that much more difficult to ask the users to bin or histogram their data before sending it to the function? In your example, `wavelength_theta_figure(da.bin(wavelength=3, theta=3)` is actually shorter than the other cases. Can we then do away with the complicated function interface?

For context, this resembles some of the functionality we had in the first version of the plotting, which would accept many types of inputs, and do different things depending on what was available. We ended up removing all that in favour of a simpler interface in Plopp and it's been much better since, I feel.

Reply:
> What happens if you do `wavelength_theta_figure(da.bin(theta=3), wavelength_bins=wavelength_bins, theta_bins=theta_bins)`?

Then the function re-bins using the provided `theta_bins`.

There are some advantages doing it this way:

1. If the input is binned in other dimensions than `wavelength` and `theta` the user has to explicitly concatenate those and then bin in the missing dimensions. If necessary that is done automatically in the current implementation. It's one less thing to think about.
2. If they write code using this function and they don't know for sure what shape their input will have, then they have to re-implement the logic handling the different cases.

Reply:
> For context, this resembles some of the functionality we had in the first version of the plotting, which would accept many types of inputs, and do different things depending on what was available. We ended up removing all that in favour of a simpler interface in Plopp and it's been much better since, I feel.

I'm not familiar with the details of the context. But I want to clarify that this is not about producing different things for different inputs. That would indeed make it difficult to reason about. In this case the output is always the same independent of the shape of the input. It's only a more forgiving interface.

Reply:
> The logic being that explicitly provided bins override implicitly provided bins.

I think such behaviour has generally lead to confusion or bugs in users' notebooks.

> If they write code using this function and they don't know for sure what shape their input will have

I'm not sure why you wouldn't know what your input is. Especially if you are using a `sciline` workflow to compute you reduced result, it will be pretty well defined and documented what you are getting at the end?

Now if you are talking about power-users computing some custom data arrays etc, then I would say that they should be able to concatenate bins and reshape to get what they want.

The fact that your function has some logic that is difficult to follow even for me, suggests that there may be an alternative way of attacking the problem?
Maybe it could be that we make these types available as outputs in the workflow, i.e. `IntensityVsThetaWavelength` or something like that.
Then we compute those and we can just call `.plot()` or maybe `.hist().plot()`.
Then the workflow would control what the inputs are, and you don't have to handle many different cases.

Reply:
Hmm, I hear you but I still think it's important to have an interface that accept different kinds of inputs as long as that makes obvious sense. Otherwise users just run into a wall as soon as they doing anything just slightly outside of the ordinary.

But let's talk more about this in person.

It's very likely we can figure out a way to simplify the implementation.